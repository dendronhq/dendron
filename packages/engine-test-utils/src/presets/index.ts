import { DVault, VaultUtils } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { WorkspaceOpts } from "../engine";

export * from "./engine-server";
export * from "./pods-core";
export * from "./vaults";

export enum SETUP_HOOK_KEYS {
  /**
   * alpha: link(beta)
   * beta: link(alpha)
   */
  WITH_LINKS = "WITH_LINKS",
}

const createLink = (fname: string, opts: { vaultPrefix?: DVault }) => {
  let cVaultPrefix = "";
  if (opts.vaultPrefix) {
    cVaultPrefix = VaultUtils.toURIPrefix(opts.vaultPrefix) + "/";
  }
  return `[[${cVaultPrefix}${fname}]]`;
};

export async function callSetupHook(
  key: SETUP_HOOK_KEYS,
  opts: {
    workspaceType: "single" | "multi";
    withVaultPrefix?: boolean;
  } & WorkspaceOpts
) {
  const { workspaceType, vaults, wsRoot, withVaultPrefix } = opts;
  const isMultiVault = workspaceType !== "single";
  const cVaults = isMultiVault ? vaults : [vaults[0], vaults[0]];

  // WITH LINKS
  if (key === SETUP_HOOK_KEYS.WITH_LINKS) {
    const link1 = createLink("beta", {
      vaultPrefix: withVaultPrefix ? cVaults[1] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "alpha",
      wsRoot,
      vault: cVaults[0],
      body: link1,
    });
    const link2 = createLink("alpha", {
      vaultPrefix: withVaultPrefix ? cVaults[0] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "beta",
      wsRoot,
      vault: cVaults[1],
      body: link2,
    });
    if (isMultiVault) {
      await NoteTestUtilsV4.createNote({
        fname: "beta",
        wsRoot,
        vault: cVaults[0],
        body: link2,
        genRandomId: true,
      });
    }
  } else {
    throw Error("not supported key");
  }
}
