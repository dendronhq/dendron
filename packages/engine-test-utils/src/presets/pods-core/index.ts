import JSON from "./json";
import MARKDOWN from "./markdown";

export const PODS_CORE = {
  JSON,
  MARKDOWN,
};

export const PODS_PRESETS = [{ name: "json", presets: JSON }];
