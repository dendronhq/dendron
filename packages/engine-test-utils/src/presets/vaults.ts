import { DVault } from "@dendronhq/common-all";

export const VAULTS = {
  MULTI_VAULT_WITH_THREE_VAULTS: () => {
    return [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ] as DVault[];
  },
};
