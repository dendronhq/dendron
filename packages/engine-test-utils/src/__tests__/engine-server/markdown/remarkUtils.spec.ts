import {
  DEngineClient,
  DLink,
  NoteProps,
  WorkspaceOpts,
} from "@dendronhq/common-all";
import { DConfig } from "@dendronhq/common-server";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import {
  DendronASTDest,
  DendronASTTypes,
  LinkFilter,
  LinkUtils,
  MDUtilsV5,
  RemarkUtils,
} from "@dendronhq/unified";
import _ from "lodash";
import { runEngineTestV5, testWithEngine } from "../../../engine";
import { ENGINE_HOOKS } from "../../../presets";
import { checkString } from "../../../utils";

const checkLink = ({ src, target }: { src: Partial<DLink>; target: DLink }) => {
  const allTrue = _.every([
    _.isUndefined(src.from) ? true : _.isEqual(src.from, target.from),
    _.isUndefined(src.type) ? true : _.isEqual(src.type, target.type),
  ]);
  if (!allTrue) {
    throw Error(
      `diff between ${JSON.stringify(src, null, 4)} and ${JSON.stringify(
        target,
        null,
        4
      )}`
    );
  }
  expect(true).toBeTruthy();
};

describe("RemarkUtils and LinkUtils", () => {
  describe("findAnchors", () => {
    test("one header", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const note = (await engine.getNote("foo")).data!;
          const body = note.body;
          const out = RemarkUtils.findAnchors(body);
          expect(out).toMatchSnapshot();
          expect(_.size(out)).toEqual(1);
          // @ts-ignore
          expect(out[0].depth).toEqual(1);
          expect(out[0].type).toEqual(DendronASTTypes.HEADING);
        },
        {
          preSetupHook: async ({ vaults, wsRoot }) => {
            await NoteTestUtilsV4.createNote({
              fname: "foo",
              body: [
                "# h1",
                "",
                "Repellendus possimus voluptates tempora quia.",
              ].join("\n"),
              vault: vaults[0],
              wsRoot,
            });
          },
          expect,
        }
      );
    });
    test("one block anchor", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const note = (await engine.getNote("foo")).data!;
          const body = note.body;
          const out = RemarkUtils.findAnchors(body);
          expect(out).toMatchSnapshot();
          expect(_.size(out)).toEqual(1);
          expect(out[0].type).toEqual(DendronASTTypes.BLOCK_ANCHOR);
        },
        {
          preSetupHook: async ({ vaults, wsRoot }) => {
            await NoteTestUtilsV4.createNote({
              fname: "foo",
              body: [
                "Repellendus possimus voluptates tempora quia.",
                "Eum deleniti sit delectus officia rem. ^block-anchor",
                "",
                "Consectetur blanditiis facilis nulla mollitia.",
              ].join("\n"),
              vault: vaults[0],
              wsRoot,
            });
          },
          expect,
        }
      );
    });
    test("doesn't find block anchor within wikilink", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const note = (await engine.getNote("foo")).data!;
          const body = note.body;
          const out = RemarkUtils.findAnchors(body);
          expect(out).toMatchSnapshot();
          expect(_.size(out)).toEqual(0);
        },
        {
          preSetupHook: async ({ vaults, wsRoot }) => {
            await NoteTestUtilsV4.createNote({
              fname: "foo",
              body: [
                "Repellendus possimus voluptates tempora quia.",
                "Eum deleniti sit delectus officia rem.",
                "",
                "[[bar#^block-anchor]]",
                "",
                "Consectetur blanditiis facilis nulla mollitia.",
              ].join("\n"),
              vault: vaults[0],
              wsRoot,
            });
          },
          expect,
        }
      );
    });
  });
  test("doesn't find block anchor within inline code", async () => {
    await runEngineTestV5(
      async ({ engine }) => {
        const note = (await engine.getNote("foo")).data!;
        const body = note.body;
        const out = RemarkUtils.findAnchors(body);
        expect(out).toMatchSnapshot();
        expect(_.size(out)).toEqual(0);
      },
      {
        preSetupHook: async ({ vaults, wsRoot }) => {
          await NoteTestUtilsV4.createNote({
            fname: "foo",
            body: [
              "Repellendus possimus voluptates tempora quia.",
              "Eum deleniti sit delectus officia rem. `^block-anchor`",
              "",
              "Consectetur blanditiis facilis nulla mollitia.",
            ].join("\n"),
            vault: vaults[0],
            wsRoot,
          });
        },
        expect,
      }
    );
  });
  test("doesn't find block anchor within code block", async () => {
    await runEngineTestV5(
      async ({ engine }) => {
        const note = (await engine.getNote("foo")).data!;
        const body = note.body;
        const out = RemarkUtils.findAnchors(body);
        expect(out).toMatchSnapshot();
        expect(_.size(out)).toEqual(0);
      },
      {
        preSetupHook: async ({ vaults, wsRoot }) => {
          await NoteTestUtilsV4.createNote({
            fname: "foo",
            body: [
              "Repellendus possimus voluptates tempora quia.",
              "Eum deleniti sit delectus officia rem.",
              "",
              "```",
              "^block-anchor",
              "```",
              "",
              "Consectetur blanditiis facilis nulla mollitia.",
            ].join("\n"),
            vault: vaults[0],
            wsRoot,
          });
        },
        expect,
      }
    );
  });

  describe("findLinksFromBody", () => {
    testWithEngine(
      "one link",
      async ({ engine, wsRoot }) => {
        const note = (await engine.getNote("foo")).data!;
        const config = DConfig.readConfigSync(wsRoot);
        const links = LinkUtils.findLinksFromBody({ note, config });
        expect(links).toMatchSnapshot();
        expect(links[0].to?.fname).toEqual("bar");
      },
      {
        preSetupHook: async ({ wsRoot, vaults }) => {
          await NoteTestUtilsV4.createNote({
            fname: "foo",
            body: "[[bar]]",
            vault: vaults[0],
            wsRoot,
          });
        },
      }
    );

    testWithEngine(
      "empty link",
      async ({ engine, wsRoot }) => {
        const note = (await engine.getNote("foo")).data!;
        const config = DConfig.readConfigSync(wsRoot);
        const links = LinkUtils.findLinksFromBody({ note, config });
        expect(links).toMatchSnapshot();
        expect(_.isEmpty(links)).toBeTruthy();
      },
      {
        preSetupHook: async ({ wsRoot, vaults }) => {
          await NoteTestUtilsV4.createNote({
            fname: "foo",
            body: "[[]]",
            vault: vaults[0],
            wsRoot,
          });
        },
      }
    );

    testWithEngine(
      "xvault link",
      async ({ engine, wsRoot }) => {
        const note = (await engine.getNote("foo")).data!;
        const config = DConfig.readConfigSync(wsRoot);
        const links = LinkUtils.findLinksFromBody({ note, config });
        expect(links).toMatchSnapshot();
        expect(links[0].from).toEqual({
          fname: "foo",
          id: "foo",
          vaultName: "vault1",
        });
        expect(links[0].to).toEqual({
          fname: "bar",
          vaultName: "vault2",
        });
      },
      {
        preSetupHook: async ({ wsRoot, vaults }) => {
          await NoteTestUtilsV4.createNote({
            fname: "foo",
            body: "[[dendron://vault2/bar]]",
            vault: vaults[0],
            wsRoot,
          });
        },
      }
    );

    testWithEngine(
      "hashtag link",
      async ({ engine, wsRoot }) => {
        const note = (await engine.getNote("foo")).data!;
        const config = DConfig.readConfigSync(wsRoot);
        const links = LinkUtils.findLinksFromBody({ note, config });
        expect(links).toMatchSnapshot();
        expect(links[0].to?.fname).toEqual("tags.bar");
      },
      {
        preSetupHook: async ({ wsRoot, vaults }) => {
          await NoteTestUtilsV4.createNote({
            fname: "foo",
            body: "#bar",
            vault: vaults[0],
            wsRoot,
          });
        },
      }
    );

    test("note ref", async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const note = (await engine.getNote("foo.one-id")).data!;
          const config = DConfig.readConfigSync(wsRoot);
          const links = LinkUtils.findLinksFromBody({ note, config });
          expect(links).toMatchSnapshot();
          checkLink({
            src: {
              from: {
                fname: "foo.one",
                id: "foo.one-id",
                vaultName: "vault1",
              },
              type: "wiki",
            },
            target: links[0],
          });
          checkLink({
            src: {
              from: {
                fname: "foo.one",
                id: "foo.one-id",
                vaultName: "vault1",
              },
              type: "ref",
            },
            target: links[1],
          });
        },
        {
          expect,
          preSetupHook: async (opts) => {
            await ENGINE_HOOKS.setupNoteRefRecursive(opts);
          },
        }
      );
    });

    describe("filter", () => {
      const preSetupHook = async ({ wsRoot, vaults }: WorkspaceOpts) => {
        await NoteTestUtilsV4.createNote({
          fname: "foo",
          wsRoot,
          vault: vaults[0],
          body: ["[[foo]]"].join("\n"),
        });
      };

      const getLinks = async (engine: DEngineClient, filter: LinkFilter) => {
        const note = (await engine.getNote("foo")).data!;
        const config = DConfig.readConfigSync(engine.wsRoot);
        const links = LinkUtils.findLinksFromBody({
          note,
          filter,
          config,
        });
        expect(links).toMatchSnapshot();
        return links;
      };

      test("loc match", async () => {
        await runEngineTestV5(
          async ({ engine }) => {
            const links = await getLinks(engine, {
              loc: { fname: "foo" },
            });
            checkLink({
              src: {
                from: {
                  fname: "foo",
                  id: "foo",
                  vaultName: "vault1",
                },
              },
              target: links[0],
            });
          },
          {
            expect,
            preSetupHook,
          }
        );
      });

      test("loc no match", async () => {
        await runEngineTestV5(
          async ({ engine }) => {
            const links = await getLinks(engine, {
              loc: { fname: "bar" },
            });
            expect(_.isEmpty(links)).toBeTruthy();
          },
          {
            expect,
            preSetupHook,
          }
        );
      });
    });
  });

  describe("updateLink", () => {
    test("basic", async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const note = (await engine.getNote("foo.one-id")).data!;
          const config = DConfig.readConfigSync(wsRoot);
          const links = LinkUtils.findLinksFromBody({ note, config });
          const link = LinkUtils.dlink2DNoteLink(links[0]);
          const newBody = LinkUtils.updateLink({
            note,
            oldLink: link,
            newLink: {
              ...link,
              from: {
                fname: "foo.bar",
              },
            },
          });
          expect(newBody).toMatchSnapshot();
          await checkString(newBody, "Regular wikilink: [[foo.bar]]");
        },
        {
          preSetupHook: async (opts) => {
            await ENGINE_HOOKS.setupNoteRefRecursive(opts);
          },
          expect,
        }
      );
    });

    describe("multiple links present", () => {
      const preSetupHook = async ({ wsRoot, vaults }: WorkspaceOpts) => {
        const vault = vaults[0];
        await NoteTestUtilsV4.createNote({
          fname: "foo",
          wsRoot,
          vault,
          body: ["[[foo]]", "nospace[[foo]]", "onespace [[foo]]"].join("\n"),
        });
      };

      test("only link", async () => {
        await runEngineTestV5(
          async ({ engine, wsRoot }) => {
            const note = (await engine.getNote("foo")).data!;
            const config = DConfig.readConfigSync(wsRoot);
            const links = LinkUtils.findLinksFromBody({ note, config });
            const link = LinkUtils.dlink2DNoteLink(links[0]);
            const newLink = {
              ...link,
              from: {
                fname: "bar",
              },
            };
            const newBody = LinkUtils.updateLink({
              note,
              oldLink: link,
              newLink,
            });
            expect(newBody).toMatchSnapshot();
            await checkString(newBody.split("\n")[0], "[[bar]]");
          },
          {
            preSetupHook,
            expect,
          }
        );
      });

      test("link no space", async () => {
        await runEngineTestV5(
          async ({ engine, wsRoot }) => {
            const idx = 1;
            const newLine = "nospace[[bar]]";
            const note = (await engine.getNote("foo")).data!;
            const config = DConfig.readConfigSync(wsRoot);
            const links = LinkUtils.findLinksFromBody({ note, config });
            const link = LinkUtils.dlink2DNoteLink(links[idx]);
            const newLink = {
              ...link,
              from: {
                fname: "bar",
              },
            };
            const newBody = LinkUtils.updateLink({
              note,
              oldLink: link,
              newLink,
            });
            expect(newBody).toMatchSnapshot();
            await checkString(newBody.split("\n")[idx], newLine);
          },
          {
            preSetupHook,
            expect,
          }
        );
      });

      test("link onespace", async () => {
        await runEngineTestV5(
          async ({ engine, wsRoot }) => {
            const idx = 2;
            const newLine = "onespace [[bar]]";
            const note = (await engine.getNote("foo")).data!;
            const config = DConfig.readConfigSync(wsRoot);
            const links = LinkUtils.findLinksFromBody({ note, config });
            const link = LinkUtils.dlink2DNoteLink(links[idx]);
            const newLink = {
              ...link,
              from: {
                fname: "bar",
              },
            };
            const newBody = LinkUtils.updateLink({
              note,
              oldLink: link,
              newLink,
            });
            expect(newBody).toMatchSnapshot();
            await checkString(newBody.split("\n")[idx], newLine);
          },
          {
            preSetupHook,
            expect,
          }
        );
      });
    });
  });

  describe("extractBlocks", () => {
    test("paragraphs", async () => {
      let note: NoteProps | undefined;
      await runEngineTestV5(
        async ({ wsRoot }) => {
          expect(note).toBeTruthy();
          const config = DConfig.readConfigSync(wsRoot);
          const blocks = await RemarkUtils.extractBlocks({
            note: note!,
            config,
          });
          expect(blocks.length).toEqual(3);
        },
        {
          expect,
          preSetupHook: async ({ wsRoot, vaults }) => {
            note = await NoteTestUtilsV4.createNote({
              wsRoot,
              vault: vaults[0],
              fname: "foo",
              body: [
                "Et et quam culpa.",
                "",
                "Cumque molestiae qui deleniti.",
                "Eius odit commodi harum.",
                "",
                "Sequi ut non delectus tempore.",
              ].join("\n"),
            });
          },
        }
      );
    });

    test("list", async () => {
      let note: NoteProps | undefined;
      await runEngineTestV5(
        async ({ wsRoot }) => {
          expect(note).toBeTruthy();
          const config = DConfig.readConfigSync(wsRoot);
          const blocks = await RemarkUtils.extractBlocks({
            note: note!,
            config,
          });
          expect(blocks.length).toEqual(5);
        },
        {
          expect,
          preSetupHook: async ({ wsRoot, vaults }) => {
            note = await NoteTestUtilsV4.createNote({
              wsRoot,
              vault: vaults[0],
              fname: "foo",
              body: [
                "Et et quam culpa.",
                "",
                "* Cumque molestiae qui deleniti.",
                "* Eius odit commodi harum.",
                "",
                "Sequi ut non delectus tempore.",
              ].join("\n"),
            });
          },
        }
      );
    });

    test("nested list", async () => {
      let note: NoteProps | undefined;
      await runEngineTestV5(
        async ({ wsRoot }) => {
          expect(note).toBeTruthy();
          const config = DConfig.readConfigSync(wsRoot);
          const blocks = await RemarkUtils.extractBlocks({
            note: note!,
            config,
          });
          expect(blocks.length).toEqual(8);
        },
        {
          expect,
          preSetupHook: async ({ wsRoot, vaults }) => {
            note = await NoteTestUtilsV4.createNote({
              wsRoot,
              vault: vaults[0],
              fname: "foo",
              body: [
                "Et et quam culpa.",
                "",
                "* Cumque molestiae qui deleniti.",
                "* Eius odit commodi harum.",
                "  * Sequi ut non delectus tempore.",
                "  * In delectus quam sunt unde.",
                "* Quasi ex debitis aut sed.",
                "",
                "Perferendis officiis ut non.",
              ].join("\n"),
            });
          },
        }
      );
    });

    test("table", async () => {
      let note: NoteProps | undefined;
      await runEngineTestV5(
        async ({ wsRoot }) => {
          expect(note).toBeTruthy();
          const config = DConfig.readConfigSync(wsRoot);
          const blocks = await RemarkUtils.extractBlocks({
            note: note!,
            config,
          });
          expect(blocks.length).toEqual(3);
        },
        {
          expect,
          preSetupHook: async ({ wsRoot, vaults }) => {
            //const txt = `# Hello Heading\nHello Content`;
            note = await NoteTestUtilsV4.createNote({
              wsRoot,
              vault: vaults[0],
              fname: "foo",
              body: [
                "Et et quam culpa.",
                "",
                "| Sapiente | accusamus |",
                "|----------|-----------|",
                "| Laborum  | libero    |",
                "| Ullam    | optio     |",
                "",
                "Sequi ut non delectus tempore.",
              ].join("\n"),
            });
          },
        }
      );
    });

    test("existing anchors", async () => {
      let note: NoteProps | undefined;
      await runEngineTestV5(
        async ({ wsRoot }) => {
          expect(note).toBeTruthy();
          const config = DConfig.readConfigSync(wsRoot);
          const blocks = await RemarkUtils.extractBlocks({
            note: note!,
            config,
          });
          expect(blocks.length).toEqual(7);
          expect(blocks[0].anchor?.value).toEqual("et-et-quam-culpa");
          expect(blocks[1].anchor?.value).toEqual("paragraph");
          expect(blocks[2].anchor?.value).toEqual("item1");
          expect(blocks[3].anchor?.value).toEqual("item2");
          expect(blocks[4].anchor?.value).toEqual("item3");
          expect(blocks[5].anchor?.value).toEqual("list");
          expect(blocks[6].anchor?.value).toEqual("table");
        },
        {
          expect,
          preSetupHook: async ({ wsRoot, vaults }) => {
            //const txt = `# Hello Heading\nHello Content`;
            note = await NoteTestUtilsV4.createNote({
              wsRoot,
              vault: vaults[0],
              fname: "foo",
              body: [
                "# Et et quam culpa. ^header",
                "",
                "Ullam vel eius reiciendis. ^paragraph",
                "",
                "* Cumque molestiae qui deleniti. ^item1",
                "* Eius odit commodi harum. ^item2",
                "  * Sequi ut non delectus tempore. ^item3",
                "",
                "^list",
                "",
                "| Sapiente | accusamus |",
                "|----------|-----------|",
                "| Laborum  | libero    |",
                "| Ullam    | optio     | ^table",
              ].join("\n"),
            });
          },
        }
      );
    });

    test("header", async () => {
      let note: NoteProps | undefined;
      await runEngineTestV5(
        async ({ wsRoot }) => {
          expect(note).toBeTruthy();
          const config = DConfig.readConfigSync(wsRoot);
          const blocks = await RemarkUtils.extractBlocks({
            note: note!,
            config,
          });
          expect(blocks.length).toEqual(4);
          expect(blocks[0].anchor?.value).toEqual("et-et-quam-culpa");
          expect(blocks[2].anchor?.value).toEqual("eius-odit-commodi-harum");
        },
        {
          expect,
          preSetupHook: async ({ wsRoot, vaults }) => {
            //const txt = `# Hello Heading\nHello Content`;
            note = await NoteTestUtilsV4.createNote({
              wsRoot,
              vault: vaults[0],
              fname: "foo",
              body: [
                "# Et et quam culpa. ^anchor",
                "",
                "Cumque molestiae qui deleniti.",
                "",
                "# Eius odit commodi harum.",
                "",
                "Sequi ut non delectus tempore.",
              ].join("\n"),
            });
          },
        }
      );
    });
  });

  describe("findLinkCandidates", () => {
    const preSetupHook = async ({ wsRoot, vaults }: WorkspaceOpts) => {
      await NoteTestUtilsV4.createNote({
        fname: "foo",
        wsRoot,
        vault: vaults[0],
        body: ["Lorem ipsum"].join("\n"),
      });

      await NoteTestUtilsV4.createNote({
        fname: "bar",
        wsRoot,
        vault: vaults[0],
        body: ["this is a test.", "see foo for more examples."].join("\n"),
      });

      await NoteTestUtilsV4.createNote({
        fname: "alpha",
        wsRoot,
        vault: vaults[0],
        body: "note that has lots of candidates from baz",
      });

      await NoteTestUtilsV4.createNote({
        fname: "baz",
        wsRoot,
        vault: vaults[0],
        body: "alpha alpha alpha lorem ipsum alpha one two three alpha",
      });

      await NoteTestUtilsV4.createNote({
        fname: "one",
        wsRoot,
        vault: vaults[0],
        body: "Lorem ipsum",
      });

      await NoteTestUtilsV4.createNote({
        fname: "two",
        wsRoot,
        vault: vaults[0],
        body: "Lorem ipsum",
      });

      await NoteTestUtilsV4.createNote({
        fname: "three",
        wsRoot,
        vault: vaults[0],
        body: "Lorem ipsum",
      });

      await NoteTestUtilsV4.createNote({
        fname: "nodes",
        wsRoot,
        vault: vaults[0],
        body: [
          "# one",
          "",
          "## one",
          "",
          "### one",
          "",
          "#### one",
          "",
          "##### one",
          "",
          "###### one",
          "",
          "> one", // allowed
          "",
          "`one`",
          "",
          "```",
          "one",
          "```",
          "",
          "- one", // allowed
          "",
          "- [ ] one", // allowed
          "",
          "- [x] one", // allowed
          "",
          "* one", // allowed
          "",
          "<div>one</div>",
          "",
          "<div>",
          "\tone",
          "</div>",
          "",
          "[one]: https://one.com",
          "",
          "*one*",
          "",
          "_one_",
          "",
          "**one**",
          "",
          "__one__",
          "",
          "[one](https://one.com)",
          "",
          '![one](https://one.com/one.ico "one")',
          "",
          "[one][stuff]",
          "",
          "[stuff][one]",
          "",
          "![one][stuff]",
          "",
          "![stuff][one]",
          "",
          "~~one~~",
          "",
          "[^one]",
          "",
          "[^one]: stuff",
          "",
          "[^stuff]: one",
          "",
          "| stuff | one   |", // allowed
          "|-------|-------|",
          "| one   | stuff |", // allowed
          "| stuff | stuff |",
          "",
        ].join("\n"),
      });
    };

    test("basic", async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const note = (await engine.getNote("bar")).data!;
          const linkCandidates = await LinkUtils.findLinkCandidates({
            note,
            engine,
            config: DConfig.readConfigSync(wsRoot),
          });
          expect(linkCandidates[0].from.fname).toEqual("bar");
          expect(linkCandidates[0].to!.fname).toEqual("foo");
          expect(linkCandidates[0].type).toEqual("linkCandidate");
        },
        {
          expect,
          preSetupHook,
        }
      );
    });

    test("multiple link candidates in one paragraph", async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const note = (await engine.getNote("baz")).data!;
          const linkCandidates = await LinkUtils.findLinkCandidates({
            note,
            engine,
            config: DConfig.readConfigSync(wsRoot),
          });
          expect(linkCandidates.length).toEqual(8);
        },
        {
          expect,
          preSetupHook,
        }
      );
    });

    test("only works for direct child of paragraph and table cell", async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const note = (await engine.getNote("nodes")).data!;
          const linkCandidates = await LinkUtils.findLinkCandidates({
            note,
            engine,
            config: DConfig.readConfigSync(wsRoot),
          });
          expect(linkCandidates.length).toEqual(8);
        },
        {
          expect,
          preSetupHook,
        }
      );
    });
  });
});

describe("h1ToTitle", () => {
  test("basic", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot, vaults }) => {
        const config = DConfig.readConfigSync(wsRoot);

        const noteToRender = await NoteTestUtilsV4.createNote({
          wsRoot,
          vault: vaults[0],
          fname: "foo",
          body: [`# Foo Header`, `## Foo Content`].join("\n"),
        });

        const proc = MDUtilsV5.procRemarkFull({
          noteToRender,
          dest: DendronASTDest.MD_REGULAR,
          fname: "foo",
          vault: vaults[0],
          config,
        });
        const engineNotes = await engine.findNotes({
          excludeStub: false,
        });
        await Promise.all(
          engineNotes.map(async (note) => {
            const newBody = await proc()
              .use(RemarkUtils.h1ToTitle(note, []))
              .process(note.body);
            note.body = newBody.toString();
            return note;
          })
        );
      },
      {
        expect,
        preSetupHook: async ({ wsRoot, vaults }) => {
          //const txt = `# Hello Heading\nHello Content`;
          await NoteTestUtilsV4.createNote({
            wsRoot,
            vault: vaults[0],
            fname: "foo",
            body: [`# Foo Header`, `## Foo Content`].join("\n"),
          });
          await NoteTestUtilsV4.createNote({
            wsRoot,
            vault: vaults[0],
            fname: "bar",
            body: [`# Bar Header`, `## Bar Content`].join("\n"),
          });
        },
      }
    );
  });
});
