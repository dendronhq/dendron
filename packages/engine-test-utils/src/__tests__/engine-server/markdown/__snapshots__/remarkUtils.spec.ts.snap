// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemarkUtils and LinkUtils doesn't find block anchor within code block 1`] = `Array []`;

exports[`RemarkUtils and LinkUtils doesn't find block anchor within inline code 1`] = `Array []`;

exports[`RemarkUtils and LinkUtils findAnchors doesn't find block anchor within wikilink 1`] = `Array []`;

exports[`RemarkUtils and LinkUtils findAnchors one block anchor 1`] = `
Array [
  Object {
    "id": "block-anchor",
    "position": Position {
      "end": Object {
        "column": 53,
        "line": 2,
        "offset": 98,
      },
      "indent": Array [],
      "start": Object {
        "column": 40,
        "line": 2,
        "offset": 85,
      },
    },
    "type": "blockAnchor",
    "value": "^block-anchor",
  },
]
`;

exports[`RemarkUtils and LinkUtils findAnchors one header 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "position": Position {
          "end": Object {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
          "indent": Array [],
          "start": Object {
            "column": 3,
            "line": 1,
            "offset": 2,
          },
        },
        "type": "text",
        "value": "h1",
      },
    ],
    "depth": 1,
    "position": Position {
      "end": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "indent": Array [],
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": "heading",
  },
]
`;

exports[`RemarkUtils and LinkUtils findLinks empty link 1`] = `Array []`;

exports[`RemarkUtils and LinkUtils findLinks filter loc match 1`] = `
Array [
  Object {
    "alias": "foo",
    "from": Object {
      "fname": "foo",
      "id": "foo",
      "vaultName": "vault1",
    },
    "position": Position {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "indent": Array [],
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "to": Object {
      "anchorHeader": undefined,
      "fname": "foo",
      "vaultName": undefined,
    },
    "type": "wiki",
    "value": "foo",
    "xvault": false,
  },
]
`;

exports[`RemarkUtils and LinkUtils findLinks filter loc no match 1`] = `Array []`;

exports[`RemarkUtils and LinkUtils findLinks note ref 1`] = `
Array [
  Object {
    "alias": "foo.two",
    "from": Object {
      "fname": "foo.one",
      "id": "foo.one-id",
      "vaultName": "vault1",
    },
    "position": Position {
      "end": Object {
        "column": 30,
        "line": 3,
        "offset": 52,
      },
      "indent": Array [],
      "start": Object {
        "column": 19,
        "line": 3,
        "offset": 41,
      },
    },
    "to": Object {
      "anchorHeader": undefined,
      "fname": "foo.two",
      "vaultName": undefined,
    },
    "type": "wiki",
    "value": "foo.two",
    "xvault": false,
  },
  Object {
    "from": Object {
      "fname": "foo.one",
      "id": "foo.one-id",
      "vaultName": "vault1",
    },
    "position": Position {
      "end": Object {
        "column": 13,
        "line": 2,
        "offset": 22,
      },
      "indent": Array [],
      "start": Object {
        "column": 1,
        "line": 2,
        "offset": 10,
      },
    },
    "to": Object {
      "anchorHeader": undefined,
      "fname": "foo.two",
      "vaultName": undefined,
    },
    "type": "ref",
    "value": "foo.two",
    "xvault": false,
  },
]
`;

exports[`RemarkUtils and LinkUtils findLinks one link 1`] = `
Array [
  Object {
    "alias": "bar",
    "from": Object {
      "fname": "foo",
      "id": "foo",
      "vaultName": "vault1",
    },
    "position": Position {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "indent": Array [],
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "to": Object {
      "anchorHeader": undefined,
      "fname": "bar",
      "vaultName": undefined,
    },
    "type": "wiki",
    "value": "bar",
    "xvault": false,
  },
]
`;

exports[`RemarkUtils and LinkUtils findLinks xvault link 1`] = `
Array [
  Object {
    "alias": "bar",
    "from": Object {
      "fname": "foo",
      "id": "foo",
      "vaultName": "vault1",
    },
    "position": Position {
      "end": Object {
        "column": 25,
        "line": 1,
        "offset": 24,
      },
      "indent": Array [],
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "to": Object {
      "anchorHeader": undefined,
      "fname": "bar",
      "vaultName": "vault2",
    },
    "type": "wiki",
    "value": "bar",
    "xvault": true,
  },
]
`;

exports[`RemarkUtils and LinkUtils findUnreferencedLinks basic 1`] = `
Array [
  Object {
    "from": Object {
      "fname": "bar",
      "id": "bar",
      "vaultName": "vault1",
    },
    "position": Position {
      "end": Object {
        "column": 27,
        "line": 2,
        "offset": 42,
      },
      "indent": Array [
        1,
      ],
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "to": Object {
      "fname": "foo",
      "vaultName": "vault1",
    },
    "type": "unreferenced",
    "value": "this is a test.
see foo for more examples.",
  },
]
`;

exports[`RemarkUtils and LinkUtils updateLink basic 1`] = `
"# Foo.One
![[foo.two]]
Regular wikilink: [[foo.bar]]
"
`;

exports[`RemarkUtils and LinkUtils updateLink multiple links present link no space 1`] = `
"[[foo]]
nospace[[bar]]
onespace [[foo]]
"
`;

exports[`RemarkUtils and LinkUtils updateLink multiple links present link onespace 1`] = `
"[[foo]]
nospace[[foo]]
onespace [[bar]]
"
`;

exports[`RemarkUtils and LinkUtils updateLink multiple links present only link 1`] = `
"[[bar]]
nospace[[foo]]
onespace [[foo]]
"
`;
